enumerable_project.rb

```diff
@@ -0,0 +1,139 @@
+ Copyright (c) Muhammasd Arslan
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return self unless block_given?
+
+    (0..length - 1).each do |i|
+      yield(self[i])
+    end
+    self
+  end
+
+  def my_each_with_index
+    return self unless block_given?
+
+    (0..length - 1).each do |i|
+      yield(self[i], i)
+    end
+  end
+
+  def my_select
+    return self unless block_given?
+
+    arr = []
+    my_each do |x|
+      arr << x if yield(x) == true
+    end
+    arr
+  end
+
+  def my_all?
+    return self unless block_given?
+
+    my_each do |x|
+      return false if yield(x) == false
+    end
+    true
+  end
+
+  def my_any?
+    return self unless block_given?
+
+    my_each do |x|
+      return true if yield(x) == true
+    end
+    false
+  end
+
+  def my_none?
+    return self unless block_given?
+
+    my_each do |x|
+      return false if yield(x) == true
+    end
+    true
+  end
+
+  def my_count(*args)
+    if args.empty?
+      return length unless block_given?
+
+      i = 0
+      my_each do |x|
+        i += 1 if yield(x) == true
+      end
+      i
+    else
+      i = 0
+      my_each do |x|
+        i += 1 if args[0] == x
+      end
+      i
+    end
+  end
+
+  def my_map
+    return self if proc.nil? && !block_given?
+
+    arr = []
+    my_each do |x|
+      val = proc.nil? ? yield(x) : proc.call(x)
+      arr << val
+    end
+    arr
+  end
```

Comment:
my_map

returns an Enumerator if no block is given:
`p [1, 2, 3].my_map` returns #<Enumerator>